File: algorithms-coursera-go/count_inversions.go
package main

func countInversion(nums []int) int {
	_, count := mergeSortInv(nums, 0)
	return count
}

func mergeSortInv(unsorted []int, invCount int) ([]int, int) {
	if len(unsorted) == 1 {
		return unsorted, invCount
	}

	n := len(unsorted)
	mid := n / 2

	left, right := unsorted[:mid], unsorted[mid:]

	var leftSorted []int
	leftSorted, invCount = mergeSortInv(left, invCount)

	var rightSorted []int
	rightSorted, invCount = mergeSortInv(right, invCount)

	return mergeInv(leftSorted, rightSorted, invCount)
}

func mergeInv(left, right []int, inv int) ([]int, int) {
	i, j := 0, 0

	var sorted []int
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			sorted = append(sorted, left[i])
			i++
		} else {
			sorted = append(sorted, right[j])
			j++
			inv += len(left) - i
		}
	}

	sorted = append(sorted, left[i:]...)
	sorted = append(sorted, right[j:]...)

	return sorted, inv
}

File: algorithms-coursera-rust/src/count_inversions.rs
pub fn count_inversion(nums: Vec<i32>) -> i32 {
    let (_, inv_count) = merge_sort(nums);

    inv_count
}

fn merge_sort(unsorted: Vec<i32>, inv_count: i32) -> (Vec<i32>, i32) {
    if unsorted.len() == 1 {
        return (unsorted, inv_count);
    }

    let n = unsorted.len();
    let mid = n / 2;

    let (left, right) = (unsorted[0..mid].to_vec(), unsorted[mid..n].to_vec());

    let (left_sorted, inv_left) = merge_sort(left, inv_count);
    let (right_sorted, inv_right) = merge_sort(right, inv_left);

    merge(left_sorted, right_sorted, inv_right)
}

fn merge(left: Vec<i32>, right: Vec<i32>, inv: i32) -> (Vec<i32>, i32) {
    let (mut i, mut j) = (0, 0);

    let mut sorted = vec![];
    while i < left.len() && j < right.len() {
        if left[i] < right[j] {
            sorted.push(left[i]);
            i += 1;
        } else {
            sorted.push(right[j]);
            j += 1;
            inv += (left.len() - i) as i32;
        }
    }

    sorted.extend(&left[i..]);
    sorted.extend(&right[j..]);

    (sorted, inv)
}
